File: TinyanLibSwift/classes/openal/OAL3DVector.m
//
//  OAL3DVector.m
//  TinyanLib
//
//  Created by Tinyan on 10/05/15.
//  Copyright 2010 __MyCompanyName__. All rights reserved.
//

#import "OAL3DVector.h"


@implementation OAL3DVector


@synthesize vec = m_vec;

-(id)init
{
	if ((self = [super init]))
	{
		m_vec = (float*)malloc(sizeof(float)*3);
		[self set:0:0:0];
	}
	return self;
}

- (id) init:(float)_x :(float)_y :(float)_z 
{
	if ((self = [super init]))
	{
		[self set:_x:_y:_z];
	}
	return self;
}


- (void) set: (float) _x : (float) _y : (float) _z 
{
	m_vec[0] = _x;
	m_vec[1] = _y;
	m_vec[2] = _z;
}

- (void) set: (OAL3DVector *) vector
{
	m_vec[0] = [vector x];
	m_vec[1] = [vector y];
	m_vec[2] = [vector z];
}

- (void) normalize
{
	float sum = 0.0;
	for (int i = 0;i <3 ;i++)
	{
		sum += m_vec[i] * m_vec[i];
	}
	
	float scale = sqrt(sum);
	if (scale > 0.0)
	{
		m_vec[0] /= scale;
		m_vec[1] /= scale;
		m_vec[2] /= scale;
	}
}



- (void) rotateZ: (float) radian
{
	float _x	= m_vec[0] * cos(radian) - m_vec[1] * sin(radian);
	float _y	= m_vec[0] * sin(radian) + m_vec[1] * cos(radian);
	m_vec[0]		= _x;
	m_vec[1]		= _y;
}


- (float) distanceFrom: (OAL3DVector *) vector
{	
	float dx = [vector x] - m_vec[0];
	float dy = [vector y] - m_vec[1];
	float dz = [vector z] - m_vec[2];
	return sqrt(dx*dx + dy*dy + dz*dz);
}

-(float)x
{
	return m_vec[0];
}
-(float)y
{
	return m_vec[1];
}
-(float)z
{
	return m_vec[2];
}

-(void)setx:(float)_x
{
	m_vec[0] = _x;
}
-(void)sety:(float)_y
{
	m_vec[1] = _y;
}
-(void)setz:(float)_z
{
	m_vec[2] = _z;
}


- (void) add: (OAL3DVector *) vector
{
	int i;
	for (i = 0; i < 3; i++) m_vec[i] += vector.vec[i];
}

- (void) sub: (OAL3DVector *) vector
{
	int i;
	for (i = 0; i < 3; i++) m_vec[i] -= vector.vec[i];
}

- (void) mul: (float) m
{
	int i;
	for (i = 0; i < 3; i++) m_vec[i] *= m;
}	

- (void) scaleTo: (float) scale
{
	[self normalize];
	[self mul: scale];
}

- (float) scale
{
	int i;
	float sum = 0.0;
	for (i = 0; i < 3; i++) sum += m_vec[i] * m_vec[i];
	return sqrt(sum);
}


- (void) dealloc
{
	free(m_vec);
}





@end

File: TinyanLibSwift/classes/openal/OAL3DVector.swift
//
//  OAL3DVector.swift
//  TinyanLibSwift
//
//  Created by たいにゃん on 2014/08/24.
//  Copyright (c) 2014年 bugnekosoft. All rights reserved.
//

import Foundation

public class OAL3DVector
{
	var m_vec : [Float] = [1.0 , 0.0 , 0.0]
	
	init()
	{
		self.set(1.0 , y:0.0 , z:0.0)
	}
	
	init(x:Float , y:Float , z:Float)
	{
		self.set(x , y:y , z:z)
	}

	func set(x:Float , y:Float , z:Float)
	{
		m_vec[0] = x
		m_vec[1] = y
		m_vec[2] = z
	}
	
	func set(vector:OAL3DVector)
	{
		m_vec[0] = vector.m_vec[0]
		m_vec[1] = vector.m_vec[1]
		m_vec[2] = vector.m_vec[2]
	}
	
	func normalize()
	{
		var sum : Float = 0.0
		for var i=0;i<3;i++
		{
			sum += m_vec[i] * m_vec[i]
		}
		var scale = sqrt(sum)
		if scale > 0.0
		{
			for var i=0;i<3;i++
			{
				m_vec[i] /= scale
			}
		}
	}
	
	func distanceFrom(vector : OAL3DVector) -> Float
	{
		var dx = vector.m_vec[0] - m_vec[0]
		var dy = vector.m_vec[1] - m_vec[1]
		var dz = vector.m_vec[2] - m_vec[2]
		
		return sqrt(dx*dx + dy*dy + dz*dz);
	}
	
	func add(vector : OAL3DVector)
	{
		for var i=0;i<3;i++
		{
			m_vec[i] += vector.m_vec[i]
			
		}
	}
	
	func sub(vector : OAL3DVector)
	{
		for var i=0;i<3;i++
		{
			m_vec[i] -= vector.m_vec[i]
		}
	}

	func mul(scale : Float)
	{
		for var i=0;i<3;i++
		{
			m_vec[i] *= scale
		}
	}
	
	func scaleTo(scale : Float)
	{
		self.normalize()
		self.mul(scale)
	}
	
	func scale() ->Float
	{
		var sum : Float = 0.0
		for var i=0;i<3;i++
		{
			sum += m_vec[i] * m_vec[i]
		}
		
		return sqrt(sum);
	}
	
	func rotateZ(radian : Float)
	{
		var x = m_vec[0] * cos(radian) - m_vec[1] * sin(radian);
		var y = m_vec[0] * sin(radian) + m_vec[1] * cos(radian);
		m_vec[0]		= x;
		m_vec[1]		= y;
	}
	

}